### PR TITLE
config_database: speed improvement, new features

### PR DESCRIPTION
**Pull request by alltilla**
### This PR introduces yet another 2 changes
---
`process parsers`
 * valid drivers are:
```
$ ./syslog-ng-cfg-db.py -c parser
Drivers of context "parser":
  add-contextual-data
  csv-parser
  date-parser
  db-parser
  geoip2
  grouping-by
  json-parser
  kv-parser
  linux-audit-parser
  map-value-pairs
  python
  snmptrapd-parser
  syslog-parser
  tags-parser
  xml
Print the options of DRIVER with `--context parser --driver DRIVER`.
```
---
`support inner_source/dst`
 * Process `diskq` and `hook-commands` inside sources or destinations.

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by alltilla** [comment]
There is 2 problems with the packaging part:
 1. It requires `python3` on the build machine.
 2. It requires `networkx` from `pip`, which can be complicated to acquire on the building machine.

I might make a `--with-config-database` configure flag, to limit the build to build machines with `python3` and `networkx`.

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: contrib/config_option_database/Makefile.am [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+	$(CFG_OPT_DB_DIR)/utils/OptionParser.py \
+	$(CFG_OPT_DB_DIR)/utils/Yacc2Graph.py \
+	$(CFG_OPT_DB_DIR)/utils/__init__.py \
+	$(CFG_OPT_DB_DIR)/utils/test/test_BisonGraph.py \
+	$(CFG_OPT_DB_DIR)/utils/test/test_OptionParser.py \
+	$(CFG_OPT_DB_DIR)/utils/test/test_Yacc2Graph.py
+
+if ENABLE_CONFIG_DATABASE
+contrib/config_option_database/.cache/options.json:
+	python3 -m pip install --user -r /requirements.txt
```
Comment (furiel) [inline contrib/config_option_database/Makefile.am]: / is not needed

File: contrib/config_option_database/Makefile.am [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+	$(CFG_OPT_DB_DIR)/__init__.py \
+	$(CFG_OPT_DB_DIR)/utils/BisonGraph.py \
+	$(CFG_OPT_DB_DIR)/utils/ConfigOptions.py \
+	$(CFG_OPT_DB_DIR)/utils/MergeYm.py \
+	$(CFG_OPT_DB_DIR)/utils/OptionParser.py \
+	$(CFG_OPT_DB_DIR)/utils/Yacc2Graph.py \
+	$(CFG_OPT_DB_DIR)/utils/__init__.py \
+	$(CFG_OPT_DB_DIR)/utils/test/test_BisonGraph.py \
+	$(CFG_OPT_DB_DIR)/utils/test/test_OptionParser.py \
+	$(CFG_OPT_DB_DIR)/utils/test/test_Yacc2Graph.py
```
Comment (furiel) [inline contrib/config_option_database/Makefile.am]: please add options.json to extra dist

File: contrib/config_option_database/Makefile.am (hunk 1 above)
Comment (furiel) [inline contrib/config_option_database/Makefile.am]: We are detecting the python interpreter during configure.. You can use `$(PYTHON)`, similarly as in `lex-rules.am`.

Comment (mitzkia) [reply contrib/config_option_database/Makefile.am]: @furiel Is it means that $(PYTHON) variable name will work for python3 interpreter also?

Comment (furiel) [reply contrib/config_option_database/Makefile.am]: There is no `PYTHON3` variable now. The version `PYTHON` can be both 2 or 3, detected by some heuristics. The detection can be forced with the `--with-python=3` option.

The gain is the `PYTHON` variable will work on distributions where the binary is just called `python` for the version3 of the interpreter.

File: configure.ac [hunk 3]
```diff
@@ -355,6 +355,10 @@ AC_ARG_ENABLE(native,
 
 AC_ARG_ENABLE(snmp-dest, [ --enable-snmp-dest Enable snmp destination (default: auto)],, enable_snmp_dest="auto")
 
+AC_ARG_ENABLE(config-database,
+	      [  --enable-config-database            Enable building config-database. (default: no)]
```
Comment (furiel) [inline configure.ac]: Please add this option to the configure line in dbld/tarball, and add the pip dependencies to the default docker image.

File: contrib/config_option_database/config_database.py [hunk 4]
```diff
@@ -81,15 +88,20 @@ def build_db():
...
-    Path.mkdir(cache_file.parents[0], exist_ok=True)
-    if rebuild or not cache_file.exists():
-        db = build_db()
-        with cache_file.open('w') as f:
-            json.dump(db, f, indent=2)
-    else:
-        with cache_file.open() as f:
+    script_dir = Path(__file__).parent
+    if script_installed:
+        with (script_dir / 'options.json').open() as f:
```
[the remark this answers is not in the document]
Comment (alltilla) [reply contrib/config_option_database/config_database.py]: `/` is the given operator for concatenating paths. It handles dirs differently on different OSs.
https://docs.python.org/3/library/pathlib.html#operators

**Comment by alltilla** [comment]
@furiel Removed the install part altogether, as it looks too complex for me to merge before the release. I will handle it in a following PR.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Pull request merged**